casper/src/main/resources/PoS.rhox

```diff
@@ -203,50 +205,72 @@ new  PoS,
             contract PoS(@"closeBlock", ackCh) = {
               new commitRewards,
                   changeEpoch,
-                  removeQuarantinedWithdrawers
+                  removeQuarantinedWithdrawers,
+                  payValidators
               in {
                 new blockDataCh, mvarProcessCh,
                     getBlockData(`rho:block:data`),
-                    newRewardsCh, newValidatorsCh,
-                    newWithdrawersCh, newBondsCh
+                    commitRewardsCh,
+                    newValidatorsCh, removeQuarantinedCh, paymentDoneCh
                 in {
                   getBlockData!(*blockDataCh) |
 
                   runMVar!(*stateCh, *mvarProcessCh, *ackCh) |
                   for(_, @blockNumber <- blockDataCh;
                       @state, mvarResultCh <- mvarProcessCh) {
-                    changeEpoch!(
-                      blockNumber,
-                      state.get("allBonds"),
-                      state.get("activeValidators"),
-                      state.get("withdrawers"),
-                      *newValidatorsCh)|
-
-                    removeQuarantinedWithdrawers!(
-                      blockNumber,
-                      state.get("allBonds"),
-                      state.get("withdrawers"),
-                      *newWithdrawersCh,
-                      *newBondsCh) |
 
                     commitRewards!(
                       state.get("pendingRewards"),
                       state.get("committedRewards"),
-                      *newRewardsCh) |
-
-                    for (@newRewards <- newRewardsCh;
-                         @newValidators <- newValidatorsCh;
-                         @newWithdrawers <- newWithdrawersCh ;
-                         @newBonds <- newBondsCh) {
-                      mvarResultCh!(
-                        {
-                          "pendingRewards":{},
-                          "committedRewards":newRewards,
-                          "activeValidators":newValidators,
-                          "withdrawers":newWithdrawers,
-                          "allBonds":newBonds
-                        },
-                        Nil)
+                      *commitRewardsCh) |
+
+                    for (@committedRewards <- commitRewardsCh) {
```

Comment:
Is there a reason that you call `changeEpoch` after `commitRewards` returns? Is this a concurrency issue or just coincidence?

Reply:
It's just coincidence. There is no dependency between. I'll add a new commit to remove this apparent dependency